### PR TITLE
Allow adding custom brands.

### PR DESCRIPTION
**Pull request by timgws**
`setSupportedBrand` allows adding a custom supported array.

### CODE REVIEW

**Comment by timgws** [comment]
`Job 108.12 errored` in travis, this looks like a problem with Composer, not with the build itself.

**Comment by frankdejonge** [comment]
This PR introduces global scope. More correct would be to be able to set custom brands, which would be set on the isntance, not on the static property. This because setting it on one instance now influences all which is a side-effect beyond instance level and will create unwanted side-effects.

**Comment by timgws** [comment]
@frankdejonge, at first I thought that this might mean removing the `$supported_cards` static property...  I have changed the `getSupportedBrands` and `setSupportedBrand` to create & access an instance copy of the `$supported_cards` array when there is a new brand added. 

This also helps backwards compatibility with any apps that expect `CreditCard::$supported_cards` to come back with a list of all the cards that are supported by the __shipping__ Omnipay library.

File: src/Omnipay/Common/CreditCard.php [hunk 1]
```diff
@@ -124,30 +157,54 @@ public function __construct($parameters = null)
...
+     * gateway supports it.
+     *
+     * Set $add_to_front to true if the key should be added to the front of the array
+     *
+     * @param  string  $name The name of the new supported brand.
+     * @param  string  $expression The regular expression to check if a card is supported.
+     * @param  bool    $add_to_front should the key be added to the front of the array?
+     * @return boolean success
+     */
+    public function setSupportedBrand($name, $expression, $add_to_front = false)
```
Comment (kayladnls) [inline src/Omnipay/Common/CreditCard.php]: This function would probably be better off named `addSupportedBrand()`, as that's what it's actually doing.

File: tests/Omnipay/Common/Message/AbstractRequestTest.php [hunk 2]
```diff
@@ -87,6 +101,17 @@ public function testAmountWithEmpty()
...
+     * @expectedException Omnipay\Common\Exception\InvalidRequestException
+     * @expectedExceptionMessage A zero amount is not allowed.
+     */
+    public function testAmountZeroNotAllowed()
+    {
+        $this->changeProtectedProperty('zeroAmountAllowed', false);
+        $this->request->setAmount('0.00');
+        $this->request->getAmount();
+    }
+
```
Comment (kayladnls) [inline tests/Omnipay/Common/Message/AbstractRequestTest.php]: This file seems unrelated to this pull request, correct?

Comment (timgws) [inline tests/Omnipay/Common/Message/AbstractRequestTest.php]: This is correct, however, when I was looking through the test results, I saw that this particular Exception being thrown as not being tested, so I added it to this branch.

**Comment by kayladnls** [comment]
I don't think I'm sold on the need for `private $user_defined_cards = null;`. What benefit is there to not just adding to the `$supported_cards` array?

File: tests/Omnipay/Common/Message/AbstractRequestTest.php (hunk 2 above)
Comment (kayladnls) [inline tests/Omnipay/Common/Message/AbstractRequestTest.php]: I'd be happy to merge that in, but it should be done in a separate PR.

**Comment by frankdejonge** [comment]
I'd also remove stuff like "prepend or append based on boolean stuff", this rarely fixes the problem at hand. If people require a certain order, get the defaults add your own, do your sorting, set the card types.

**Comment by timgws** [comment]
@kayladnls `$supported_brands` is a static property, check out @frankdejonge's comment from yesterday. Could we then make `$supported_brands` private?

**Comment by kayladnls** [comment]
agree, @frankdejonge. If it's a really common use case, I could see something like `shiftSupportedBrand()` and `pushSupportedBrand()`, but even that is a bit meh.

**Comment by kayladnls** [comment]
@timgws yes, I would suggest making `$supported_brands` private, and getting rid of `$user_defined_cards`

**Comment by timgws** [comment]
Guys, I have added three pull requests, please check those PRs instead.
